torch/csrc/api/include/torch/data/dataloader.h

```diff
@@ -0,0 +1,301 @@
+#pragma once
+
+#include <torch/data/data_loader_options.h>
+#include <torch/data/detail/data_shuttle.h>
+#include <torch/data/detail/sequencers.h>
+#include <torch/data/iterator.h>
+#include <torch/data/samplers/random.h>
+#include <torch/data/worker_exception.h>
+
+#include <torch/csrc/utils/memory.h>
+#include <torch/csrc/utils/variadic.h>
+
+#include <ATen/Error.h>
+#include <ATen/optional.h>
+
+#include <cstddef>
+#include <exception>
+#include <memory>
+#include <thread>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+namespace torch {
+namespace data {
+template <typename Dataset, typename Sampler>
+class DataLoader {
+ public:
+  using Batch = typename Dataset::BatchType;
+  using IndexBatch = std::vector<size_t>;
+
+  /// Constructs a new `DataLoader` from a `dataset` to sample from, `options`
+  /// to configure the `DataLoader` with, and a `sampler` that specifies the
+  /// sampling strategy.
+  DataLoader(Dataset dataset, DataLoaderOptions options, Sampler sampler)
+      : options_(std::move(options)),
+        dataset_size_(dataset.size()),
+        sampler_(std::move(sampler)),
+        sequencer_(new_sequencer()) {
+    // clang-format off
+    AT_CHECK(
+        options_.batch_size <= dataset.size(),
+        "Batch size (was ", options_.batch_size, ") ",
+        "must not be larger than the dataset size (was ",
+        dataset.size(), ")");
+    // clang-format on
+
+    if (options_.workers > 0) {
+      for (size_t w = 0; w < options_.workers; ++w) {
+        // Here we copy the dataset into the worker thread closure. Each worker
+        // has its own copy of the dataset. This means the dataset must be
+        // trivially copiable, or else we don't expect more than one worker to
+        // be in use.
+        workers_.emplace_back(
+            [this, dataset] { this->worker_thread(std::move(dataset)); });
+      }
+    } else {
+      main_thread_dataset_ = torch::make_unique<Dataset>(std::move(dataset));
+    }
+  }
+
+  ~DataLoader() {
+    join();
+  }
+
+  /// Returns an iterator into the `DataLoader`. The lifetime of the iterator is
+  /// bound to the `DataLoader`. In C++ standards language, the category of the
+  /// iterator is `OutputIterator`. See
+  /// https://en.cppreference.com/w/cpp/named_req/OutputIterator for what this
+  /// means. In short: you may increment the iterator and dereference it, but
+  /// cannot go back, or step forward more than one position at a time. When the
+  /// `DataLoader` is exhausted, it will compare equal with the special
+  /// "sentinel" iterator returned by `DataLoader::end()`. Most of the time, you
+  /// should only use range-for loops to loop over the `DataLoader`, but
+  /// standard algorithms like `std::copy(dataloader.begin(), dataloader.end(),
+  /// output_iterator)`  are supported too.
+  Iterator<Batch> begin() {
+    reset();
```

Comment:
Can we assert that we've finished the last iteration? As I mentioned in one of the comments that are still open, it works differently than Python data loaders which can be safely used in parallel. Those iterators have side effects reaching as far as reconfiguring the whole internal state of this struct.

Reply:
I've added a check. It's true there's still a difference between iterators doing all the work like in Python.